File: discord-bot-fleet/src/tools/schedule.ts
// schedule_reminder — fire-and-forget reminders using node-cron.
// Reminders fire as Discord messages from this bot to a target channel.

import cron from 'node-cron';
import type { LLMTool } from '../llm.js';
import type { ToolContext } from './index.js';

// In-process scheduler (resets on restart — acceptable for v1).
const scheduledJobs = new Map<string, cron.ScheduledTask>();

function parseDelayToCron(delay: string): { cron: string; oneShot: boolean } | null {
  // Supported formats: "5m", "30m", "1h", "2h", "1d", or a cron expression
  const m = delay.match(/^(\d+)([smhd])$/);
  if (m) {
    const n = parseInt(m[1], 10);
    const unit = m[2];
    if (unit === 's' && n >= 30) {
      // seconds -> convert to cron "*/n * * * * *" but node-cron doesn't support seconds reliably; use setTimeout instead
      return null;
    }
    if (unit === 'm') return { cron: `*/${n} * * * *`, oneShot: true };
    if (unit === 'h') return { cron: `0 */${n} * * *`, oneShot: true };
    if (unit === 'd') return { cron: `0 0 */${n} * *`, oneShot: true };
  }
  // Try as a cron expression
  if (cron.validate(delay)) {
    return { cron: delay, oneShot: false };
  }
  return null;
}

export const scheduleReminderTool: LLMTool = {
  type: 'function',
  function: {
    name: 'schedule_reminder',
    description: 'Schedule a future reminder that posts a message to a channel. Useful for "remind me in 1 hour" or scheduled pings. Delay can be "30m", "1h", "2h", "1d", or a cron expression like "0 9 * * *".',
    parameters: {
      type: 'object',
      properties: {
        delay: { type: 'string', description: 'Time until reminder fires. Format: "30m" / "1h" / "2h" / "1d", or a cron expression like "0 9 * * *".' },
        message: { type: 'string', description: 'Message content to post when the reminder fires.' },
        channel_id: { type: 'string', description: 'Discord channel ID to post in. If omitted, uses the channel the conversation is happening in.' },
      },
      required: ['delay', 'message'],
    },
  },
};

export async function scheduleReminderHandler(
  args: { delay: string; message: string; channel_id?: string },
  ctx: ToolContext
): Promise<string> {
  const parsed = parseDelayToCron(args.delay);
  if (!parsed) {
    return JSON.stringify({ error: `invalid delay "${args.delay}". Use "30m", "1h", "2h", "1d", or a cron expression.` });
  }
  const channelId = args.channel_id || ctx.botConfig.channel_ids[0];
  if (!channelId) {
    return JSON.stringify({ error: 'no channel_id provided and bot has no default channel configured' });
  }
  const jobId = `reminder_${Date.now()}_${Math.random().toString(36).slice(2, 7)}`;
  const task = cron.schedule(parsed.cron, async () => {
    try {
      await ctx.sendChannelMessage(channelId, args.message);
    } catch (e) {
      console.error('[schedule] failed to send reminder:', e);
    }
    if (parsed.oneShot) {
      task.stop();
      scheduledJobs.delete(jobId);
    }
  });
  scheduledJobs.set(jobId, task);
  return JSON.stringify({ ok: true, job_id: jobId, fires_at_cron: parsed.cron, one_shot: parsed.oneShot });
}

File: discord-bot-fleet/bot-python/tools/schedule.py
"""schedule_reminder — fire-and-forget reminders using asyncio tasks."""

import asyncio
import json
import re

SCHEDULE_TOOL = {
    'type': 'function',
    'function': {
        'name': 'schedule_reminder',
        'description': 'Schedule a future reminder that posts a message to a channel. Useful for "remind me in 1 hour" or scheduled pings. Delay can be "30m", "1h", "2h", "1d".',
        'parameters': {
            'type': 'object',
            'properties': {
                'delay': {'type': 'string', 'description': 'Time until reminder fires. Format: "30m" / "1h" / "2h" / "1d".'},
                'message': {'type': 'string', 'description': 'Message content to post when the reminder fires.'},
                'channel_id': {'type': 'string', 'description': 'Discord channel ID to post in. If omitted, uses the channel the conversation is happening in.'},
            },
            'required': ['delay', 'message'],
        },
    },
}


def parse_delay(delay: str):
    """Returns delay in seconds, or None if invalid."""
    m = re.match(r'^(\d+)([smhd])$', delay.strip().lower())
    if m:
        n = int(m.group(1))
        unit = m.group(2)
        if unit == 's':
            return n
        if unit == 'm':
            return n * 60
        if unit == 'h':
            return n * 3600
        if unit == 'd':
            return n * 86400
    return None


async def schedule_handler(args, ctx):
    delay_str = args.get('delay', '').strip()
    message = args.get('message', '').strip()
    channel_id = args.get('channel_id') or (ctx.bot_config.get('channel_ids') or [None])[0]
    if not delay_str or not message:
        return json.dumps({'error': 'delay and message are required'})
    seconds = parse_delay(delay_str)
    if seconds is None:
        return json.dumps({'error': f'invalid delay "{delay_str}". Use "30m", "1h", "2h", "1d".'})
    if not channel_id:
        return json.dumps({'error': 'no channel_id provided and bot has no default channel configured'})
    if not ctx.send_channel_message:
        return json.dumps({'error': 'no send_channel_message available in context'})

    async def fire():
        await asyncio.sleep(seconds)
        try:
            await ctx.send_channel_message(channel_id, message)
        except Exception as e:
            print(f'[schedule] failed to send reminder: {e}', flush=True)

    # Schedule but don't await
    asyncio.create_task(fire())
    job_id = f'reminder_{int(asyncio.get_event_loop().time() * 1000)}_{id(fire)}'
    return json.dumps({
        'ok': True,
        'job_id': job_id,
        'delay_seconds': seconds,
        'channel_id': channel_id,
    })
